README.md
# Continuous Deployment
GitOps Continuous deployment for *fabric-es* and its projects.


```text
Dear Everyone,

All steps are not future-proof, any change may break. It works, maybe I don't know how it works.

Take your own risk  🎃
```

### Directory Structure
**CD Application**
- app-of-app: ArgoCD application-of-application helm chart
- argo-app: ArgoCD application helm chart
- argo-wf: Argo WorkflowTemplate helm chart
- argocd: Deployment manifests for *ArgoCD* server

**Hyperledger Application**
- chaincode: for building chaincode docker image
- gupload: grpc file uploader helm chart
- hlf-ca: Hyperledger Fabric CA server helm chart
- hlf-cc: Hyperledger Fabric Chaincode helm chart
- hlf-ord: Hyperledger Fabric Orderer helm chart
- hlf-peer: Hyperledger Fabric Peer helm chart
- networking: Istio manifests
- orgadmin: cli for organization administrator
- workflow: Argo Workflow manifest for first-time bootraping

### Usable Release: dev-0.1
Quickstart deployment

- Target branch: `dev-0.1`
- Topology: 2-org;1-peer Fabric-only (org0/org1/org2)
- namespace: n0/n1/n2
- Configtx: Standard
- GCP project: fdi-cd
- GKE: dev-core-b
- GCS bucket: fabric-cd-dev
- GCP KMS: projects/fdi-cd/locations/us-central1/keyRings/fdi/cryptoKeys/sops-key

**Requirements**

- All applications are deployed to project *fdi-cd* of GKE account *hktfp.5.gmail.com*.
- All `secrets.*.yaml` are encrypted with GCP KMS key *sops-key* of keyring *fdi*. Encrypt/decrypt are made via service accounts.
- *rtang03* is only deployer. All ArgoCD application deployment and synchronization requires github SSH key of *rtang03*.

**List of deployment manifest**

- argo/*
- argocd/*
- networking/*
- workflow/*
- scripts/env.*.sh

**List of application value files**

- .sops.yaml and */.sops.yaml
- app-of-app/values-*.yaml
- argo-wf/values-*.yaml
- gupload/values-*.yaml
- hlf-ca/values-*.yaml
- hlf-ca/secrets.*.yaml
- hlf-cc/values-*.yaml
- hlf-ord/values-*.yaml
- hlf-peer/values-*.yaml
- orgadmin/secrets.*.yaml
- orgadmin/values-*.yaml

**Update Host Alias in values files**
In GCP, istio is in-cluster deployment, newly created GKE cluster will have different istio gateway ip address.
Every newly created cluster, you need to update host alias in values files.

```shell script
# obtain ingressgateway ip address
kubectl -n istio-system get svc | grep ingressgateway
```

Update ip address of below values files:
- hlf-peer/values-p0o1.yaml
- hlf-peer/values-p0o2.yaml
- hlf-ord/values-o0.yaml
- hlf-ord/values-o1.yaml
- hlf-ord/values-o2.yaml
- hlf-ord/values-o3.yaml
- hlf-ord/values-o4.yaml
- workflow/bootstrap/values.yaml


### Getting Started: first-time network setup

**Checklist**:

1. MUST READ [DEVELOPMENT](https://github.com/rtang03/fabric-cd/doc/DEVELOPMENT.md)
1. *GKE* cluster creation
1. *Argo* and *ArgoCD* installation on GKE in-cluster
1. GCP Cloud DNS, KMS, Storage setup
1. Remove pre-existing GCS storage bucket `fabric-cd-dev`. The workflows output artifacts to bucket. The non-empty paths will give error when outputing.

In local machine:
- login with *gcloud*, installed with *kubectl*, *gsutil* cli
- install SOPS, *Argo* and *ArgoCD* CLI
- activate *Argo* and *ArgoCD* port-forwarding
- update `/etc/hosts` to point to *Argo* and *ArgoCD*, the IP comes from Istio gateway

```text
# /etc/hosts
35.202.107.80 argocd.server
35.202.107.80 argo.server
```

### 1. Prepare secrets file
Credentials, secrets, and passwords are re-located to `secrets.*.yaml`, in corresponding Helm chart directories. For example,
see `orgadmin/secrets.admin1-example.yaml`, `tlsca_caadmin` must be a base64 encoded value, and created as k8s Secret resource.

```shell script
# encode ==> dGxzY2ExLWFkbWluCg==
echo -n 'tlsca1-admin' | base64

# decode => tlsca1-admin
echo -n 'dGxzY2ExLWFkbWluCg==' | base64 -d
```

`orgadmin/secrets.admin1-example.yaml` is an encoded yaml. You copy its file content into `orgadmin/secrets.admin1.yaml`;
and then run below command to perform sops encryption. `-i` means in-place replacement; the previous unencrypted/encoded
yaml will be replaced. Repeat the same steps for every `secrets.*.yaml`. The yaml property ending with "_unencrypted" will skip encryption.

```shell script
# encrypt
sops -e -i orgadmin/secrets.admin1.yaml

# decrypt
sops -d orgadmin/secrets.admin1.yaml
```

### 2. Install fabric-network - org0 and org1

The run may take 30+ minutes. In addition to CLI, you may also use GKE dashboard, *argocd* and *argo* web UI to monitor the live
status.

```shell script
cd scripts

# Optionally, if there is running application from previous installation
# this will uninstall all ArgoCD applications and workflows in all namespace.
./uninstall.argo.sh

# First time setup, will remove pre-existing PVC for org0 and org1; and create new ones
# ALL DATA WILL BE REMOVED
./recreate-pvc.sh org1

# Similarly for org2
./recreate-pvc.sh org2

# Bootstrap Org0 and Org1
./bootstrap.org1.sh
```

The successful deployment should show:

```text
Name:                bootstrap-channel-org1
Namespace:           n1
ServiceAccount:      workflow
Status:              Succeeded
Conditions:
 Completed           True
Created:             Tue Dec 22 14:11:15 +0800 (5 minutes ago)
Started:             Tue Dec 22 14:11:15 +0800 (5 minutes ago)
Finished:            Tue Dec 22 14:16:22 +0800 (now)
Duration:            5 minutes 7 seconds
ResourcesDuration:   10m54s*(1 cpu),10m54s*(100Mi memory)

STEP                              TEMPLATE                                DURATION
 ✔ bootstrap-channel-org1         main
 ├-·-✔ sync-g1                    argocd-cli/argocd-app-sync              1m
 | └-✔ sync-p0o1                  argocd-cli/argocd-app-sync              1m
 ├---✔ dl-create-tlscacert        download-and-create-secret/main
 |   ├---✔ retrieve               retrieve-tmpl                           6s
 |   ├---✔ delete-secret-tmpl     secret-resource/delete-secret-tmpl      1s
 |   └---✔ create-secret-tmpl     secret-resource/create-secret-1key-tmpl 2s
 ├---✔ create-channel(0)          create-channel/main                     11s
 ├---✔ join-channel(0)            join-channel/main                       20s
 ├---✔ update-anchor-peer         update-anchor-peer/main                 18s
 ├---✔ package-install-chaincode  package-install-chaincode/main          13s
 ├---✔ chaincode-id-resource      chaincode-id-resource/main
 |   ├---✔ delete-ccid            delete-ccid                             2s
 |   └---✔ create-ccid            create-ccid                             2s
 ├---✔ sync-chaincode             argocd-cli/argocd-app-sync              1m
 ├---✔ approve-chaincode          approve-chaincode/main                  11s
 ├---✔ commit-chaincode           commit-chaincode/main                   13s
 └---✔ smoke-test(0)              smoke-test/main                         21s
```

### 3. Install fabric-network - org2

```shell
# Bootstrap Org2
./bootstrap.orgx.sh org2
```

Successful deployment should show something below.
```shell script
Name:                bootstrap-ch-org2-mrzz8
Namespace:           n2
ServiceAccount:      workflow
Status:              Succeeded
Conditions:
 Completed           True
ResourcesDuration:   8m18s*(1 cpu),8m18s*(100Mi memory)

STEP                                  TEMPLATE                                  DURATION
 ✔ bootstrap-ch-org2-mrzz8            main
 ├-·-✔ load-org0tlscacert             download-and-create-secret/main
 | | ├---✔ retrieve                   retrieve-tmpl                             28s
 | | ├---✔ delete-secret-tmpl         secret-resource/delete-secret-tmpl        3s
 | | └---✔ create-secret-tmpl         secret-resource/create-secret-1key-tmpl   2s
 | ├-✔ load-org1tlscacert             download-and-create-secret/main
 | | ├---✔ retrieve                   retrieve-tmpl                             24s
 | | ├---✔ delete-secret-tmpl         secret-resource/delete-secret-tmpl        2s
 | | └---✔ create-secret-tmpl         secret-resource/create-secret-1key-tmpl   3s
 | └-✔ load-org2tlscacert             download-and-create-secret/main
 |   ├---✔ retrieve                   retrieve-tmpl                             26s
 |   ├---✔ delete-secret-tmpl         secret-resource/delete-secret-tmpl        4s
 |   └---✔ create-secret-tmpl         secret-resource/create-secret-1key-tmpl   2s
 ├---✔ delete-files                   gupload-up-file/delete-files-tmpl         5s
 ├---✔ curl-pull-tlscacert            curl-event/curl-tmpl                      4s
 ├-·-✔ sync-g2                        argocd-cli/argocd-app-sync                28s
 | └-✔ sync-p0o2                      argocd-cli/argocd-app-sync                1m
 ├---✔ curl-fetch-block               curl-event/curl-tmpl                      3s
 ├---✔ wait-1                         utility/sleep                             33s
 ├---✔ check-fetchconfig-log-exist    gupload-up-file/file-exist-and-no-error   5s
 ├---✔ neworg-config-update           neworg-config-update/main
 |   ├---✔ neworg                     neworg-tmpl                               23s
 |   └---✔ gupload                    gupload-up-file/upload-tmpl               7s
 ├---✔ curl-update-channel            curl-event/curl-tmpl                      3s
 ├---✔ wait-2                         utility/sleep                             33s
 ├---✔ check-updatechannel-log-exist  gupload-up-file/file-exist-and-no-error   4s
 ├---✔ join-channel-orgx              join-channel-orgx/main
 |   ├---✔ fetch-block                fetch-tmpl                                8s
 |   └---✔ join-channel               join-channel/main                         11s
 ├---✔ update-anchor-peer             update-anchor-peer/main                   19s
 ├---✔ package-install-chaincode      package-install-chaincode/main            15s
 ├---✔ chaincode-id-resource          chaincode-id-resource/main
 |   ├---✔ delete-ccid                delete-ccid                               2s
 |   └---✔ create-ccid                create-ccid                               2s
 ├---✔ sync-chaincode                 argocd-cli/argocd-app-sync                15s
 ├---✔ approve-chaincode              approve-chaincode/main                    13s
 └---✔ smoke-test                     smoke-test/main                           13s
```


### 4. Install redis, auth-server, gw-orgX, ui-control - org0 and org1

**Create tls secret to enable secure istio gateway**

For org1, both *ui-control* and *gw-org* are required to expose to public internet. Here utilizes the secure gateway of
istio (see [istio v1.6 doc](https://istio.io/v1.6/docs/tasks/traffic-management/ingress/secure-ingress/)). As below code,
`https://web.org1.net` is exposed, and requiring the secret `peer0.org1.net-tls`.

```yaml
# networking/istio-n1.yaml
- port:
    number: 443
    name: https
    protocol: HTTPS
  tls:
    mode: SIMPLE
    credentialName: "peer0.org1.net-tls"
  hosts:
    - "web.org1.net"
    - "gw.org1.net"
```

```shell
# delete existing tls cert
kubectl -n istio-system delete secret peer0.org1.net-tls

# retrieve tls cert from org1
CERT="$(kubectl -n n1 get secret peer0.org1.net-tls -o=jsonpath='{.data.tls\.crt}' | base64 --decode)"
KEY="$(kubectl -n n1 get secret peer0.org1.net-tls -o=jsonpath='{.data.tls\.key}' | base64 --decode)"

# create tls cert for Istio secure gateway
kubectl -n istio-system create secret generic peer0.org1.net-tls --from-literal=tls.crt="$CERT" --from-literal=tls.key="$KEY"

# Debugging step
kubectl logs -n istio-system "$(kubectl get pod -l istio=ingressgateway -n istio-system -o jsonpath='{.items[0].metadata.name}')"
# should return
# 2020-12-22T08:23:04.850277Z	info	Channel Connectivity change to READY
# 2020-12-22T08:45:28.961386Z	info	sds	resource:peer0.org1.net-tls pushed key/cert pair to proxy
# 2020-12-22T08:45:28.961426Z	info	sds	Dynamic push for secret peer0.org1.net-tls
```
**Synchronize apps**

```shell
argo submit -n n1 workflow/aoa-sync-re-au-gw-ui.n1.yaml --watch --request-timeout 900s
```

It should return:

```text
Name:                aoa-sync-gw-org1
Namespace:           n1
ServiceAccount:      workflow
Status:              Succeeded
Conditions:
 Completed           True
Created:             Tue Dec 22 14:39:24 +0800 (8 minutes ago)
Started:             Tue Dec 22 14:39:24 +0800 (8 minutes ago)
Finished:            Tue Dec 22 14:48:03 +0800 (now)
Duration:            8 minutes 39 seconds
ResourcesDuration:   16m57s*(1 cpu),16m57s*(100Mi memory)

STEP                 TEMPLATE                    PODNAME                      DURATION  MESSAGE
 ✔ aoa-sync-gw-org1  main
 ├---✔ sync-redis1   argocd-cli/argocd-app-sync  aoa-sync-gw-org1-3775298967  1m
 ├---✔ sync-auth1    argocd-cli/argocd-app-sync  aoa-sync-gw-org1-3067430853  1m
 ├---✔ sync-gw-org1  argocd-cli/argocd-app-sync  aoa-sync-gw-org1-1951236881  3m
 └---✔ sync-ui-org1  argocd-cli/argocd-app-sync  aoa-sync-gw-org1-2916683662  2m
```


### Tear-down
```shell script
cd scripts

./uninstall.argo.sh

./recreate-pvc.sh org1

./recreate-pvc.sh org2
```

Lastly, remove the 'fabric-cd-dev' storage bucket.




### For gitOps Contributors
Please see [DEVELOPMENT](https://github.com/rtang03/fabric-cd/doc/DEVELOPMENT.md)



CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/rtang03/fabric-cd/tree/HEAD)

[Full Changelog](https://github.com/rtang03/fabric-cd/compare/v0.1.0-dev...HEAD)

**Implemented enhancements:**

- helm chart for RediSearch [\#64](https://github.com/rtang03/fabric-cd/issues/64)
- improve documentation [\#54](https://github.com/rtang03/fabric-cd/issues/54)
- attempt Fluent bit [\#51](https://github.com/rtang03/fabric-cd/issues/51)
- Redis Helm Chart up and running [\#66](https://github.com/rtang03/fabric-cd/pull/66) ([rtang03](https://github.com/rtang03))

**Closed issues:**

- add hlf explorer [\#5](https://github.com/rtang03/fabric-cd/issues/5)

**Merged pull requests:**

- feat: add redis1, auth-server1, ui-org1, gw-org1 [\#68](https://github.com/rtang03/fabric-cd/pull/68) ([rtang03](https://github.com/rtang03))
- merge interim changes [\#67](https://github.com/rtang03/fabric-cd/pull/67) ([rtang03](https://github.com/rtang03))
- fix: misplaced files [\#63](https://github.com/rtang03/fabric-cd/pull/63) ([rtang03](https://github.com/rtang03))

## [v0.1.0-dev](https://github.com/rtang03/fabric-cd/tree/v0.1.0-dev) (2020-11-23)

[Full Changelog](https://github.com/rtang03/fabric-cd/compare/f1276fc57313c5392dabf11c437a61fc1831a011...v0.1.0-dev)

**Implemented enhancements:**

- change to automatic change log via Github Action [\#52](https://github.com/rtang03/fabric-cd/issues/52)
- expose argo workflow API Server [\#42](https://github.com/rtang03/fabric-cd/issues/42)
- integrate ArgoCD with helm-secrets and GCP [\#37](https://github.com/rtang03/fabric-cd/issues/37)
- attempt ArgoCD [\#9](https://github.com/rtang03/fabric-cd/issues/9)
- evaluate how to transit to operator pattern [\#6](https://github.com/rtang03/fabric-cd/issues/6)
- add helm tests [\#4](https://github.com/rtang03/fabric-cd/issues/4)
- install org3; require org1, org2 signing to join channel [\#2](https://github.com/rtang03/fabric-cd/issues/2)
- install org2 chaincode [\#1](https://github.com/rtang03/fabric-cd/issues/1)
- Update master with argo workflow [\#44](https://github.com/rtang03/fabric-cd/pull/44) ([rtang03](https://github.com/rtang03))
- install hyperledger component [\#10](https://github.com/rtang03/fabric-cd/pull/10) ([rtang03](https://github.com/rtang03))
- improved out of band process [\#7](https://github.com/rtang03/fabric-cd/pull/7) ([rtang03](https://github.com/rtang03))
- org2 smoke test passes [\#3](https://github.com/rtang03/fabric-cd/pull/3) ([rtang03](https://github.com/rtang03))

**Fixed bugs:**

- secrets workflow [\#46](https://github.com/rtang03/fabric-cd/issues/46)

**Closed issues:**

- add peer1 and peerX [\#49](https://github.com/rtang03/fabric-cd/issues/49)

**Merged pull requests:**

- test: 4th attempt Changelog.yaml [\#62](https://github.com/rtang03/fabric-cd/pull/62) ([rtang03](https://github.com/rtang03))
- test: 3rd attempt Changelog.yaml [\#61](https://github.com/rtang03/fabric-cd/pull/61) ([rtang03](https://github.com/rtang03))
- test: Changelog.yaml [\#60](https://github.com/rtang03/fabric-cd/pull/60) ([rtang03](https://github.com/rtang03))
- test: Changelog.yml [\#59](https://github.com/rtang03/fabric-cd/pull/59) ([rtang03](https://github.com/rtang03))
- feat: Releasing dev-0.1 [\#50](https://github.com/rtang03/fabric-cd/pull/50) ([rtang03](https://github.com/rtang03))
- merge to master [\#45](https://github.com/rtang03/fabric-cd/pull/45) ([rtang03](https://github.com/rtang03))
- clean up [\#39](https://github.com/rtang03/fabric-cd/pull/39) ([rtang03](https://github.com/rtang03))
- debug remove echo [\#33](https://github.com/rtang03/fabric-cd/pull/33) ([rtang03](https://github.com/rtang03))
- first deploy via argo cd [\#12](https://github.com/rtang03/fabric-cd/pull/12) ([rtang03](https://github.com/rtang03))



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


bootstrap.gcp.sh
. ./scripts/setup.sh

SECONDS=0

echo "#################################"
echo "### Step 0: remove secret: n0, n1"
echo "#################################"
./scripts/rm-secret.n0.sh
./scripts/rm-secret.n1.sh

echo "#################################"
echo "### Step 1: Install Admin1"
echo "#################################"
helm install admin1 -n n1 -f ./releases/org1/admin1-orgadmin.gcp.yaml ./orgadmin
printMessage "install admin1" $?

set -x
kubectl wait --for=condition=Available --timeout 180s deployment/admin1-orgadmin-cli -n n1
res=$?
set +x
printMessage "deployment/admin1-orgadmin-cli" $res

set -x
export POD_PSQL1=$(kubectl get pods -n n1 -l "app.kubernetes.io/name=postgresql-0,app.kubernetes.io/instance=admin1" -o jsonpath="{.items[0].metadata.name}")
kubectl wait --for=condition=Ready --timeout 180s pod/$POD_PSQL1 -n n1
res=$?
set +x
printMessage "pod/$POD_PSQL1" $res

sleep 30

echo "#################################"
echo "### Step 2: Install tlsca1"
echo "#################################"
helm install tlsca1 -n n1 -f ./releases/org1/tlsca1-hlf-ca.gcp.yaml ./hlf-ca
printMessage "install tlsca1" $?

set -x
export POD_TLSCA1=$(kubectl get pods -n n1 -l "app=hlf-ca,release=tlsca1" -o jsonpath="{.items[0].metadata.name}")
kubectl wait --for=condition=Ready --timeout 180s pod/$POD_TLSCA1 -n n1
res=$?
set +x
printMessage "pod/$POD_TLSCA1" $res

echo "#################################"
echo "### Step 3: Install rca1"
echo "#################################"
helm install rca1 -n n1 -f ./releases/org1/rca1-hlf-ca.gcp.yaml ./hlf-ca
printMessage "install rca1" $?

set -x
export POD_RCA1=$(kubectl get pods -n n1 -l "app=hlf-ca,release=rca1" -o jsonpath="{.items[0].metadata.name}")
kubectl wait --for=condition=Ready --timeout 180s pod/$POD_RCA1 -n n1
res=$?
set +x
printMessage "pod/$POD_RCA1" $res

sleep 30

echo "#################################"
echo "### Step 4: Job: crypto-tlsca1"
echo "#################################"
helm install crypto-tlsca1 -n n1 -f ./releases/org1/tlsca1-cryptogen.gcp.yaml ./cryptogen
printMessage "install crypto-tlsca1" $?

set -x
kubectl wait --for=condition=complete --timeout 180s job/crypto-tlsca1-cryptogen -n n1
res=$?
set +x
printMessage "job/crypto-tlsca1-cryptogen" $res

sleep 30

echo "#################################"
echo "### Step 5: Job crypto-rca1"
echo "#################################"
helm install crypto-rca1 -n n1 -f ./releases/org1/rca1-cryptogen.gcp.yaml ./cryptogen
printMessage "install crypto-rca1" $?

set -x
kubectl wait --for=condition=complete --timeout 180s job/crypto-rca1-cryptogen -n n1
res=$?
set +x
printMessage "job/crypto-rca1-cryptogen" $res

echo "#################################"
echo "### Step 6: Install admin0"
echo "#################################"
helm install admin0 -n n0 -f ./releases/org0/admin0-orgadmin.gcp.yaml ./orgadmin
printMessage "install admin0" $?

set -x
export POD_PSQL0=$(kubectl get pods -n n0 -l "app.kubernetes.io/name=postgresql-0,app.kubernetes.io/instance=admin0" -o jsonpath="{.items[0].metadata.name}")
kubectl wait --for=condition=Ready --timeout 180s pod/$POD_PSQL0 -n n0
res=$?
set +x
printMessage "pod/$POD_PSQL0" $res

sleep 30

echo "#################################"
echo "### Step 7: Install tlsca0"
echo "#################################"
helm install tlsca0 -n n0 -f ./releases/org0/tlsca0-hlf-ca.gcp.yaml ./hlf-ca
printMessage "install tlsca0" $?

set -x
export POD_TLSCA0=$(kubectl get pods -n n0 -l "app=hlf-ca,release=tlsca0" -o jsonpath="{.items[0].metadata.name}")
kubectl wait --for=condition=Ready --timeout 180s pod/$POD_TLSCA0 -n n0
res=$?
set +x
printMessage "pod/$POD_TLSCA0" $res

sleep 30

echo "#################################"
echo "### Step 8: Install rca0"
echo "#################################"
helm install rca0 -n n0 -f ./releases/org0/rca0-hlf-ca.gcp.yaml ./hlf-ca
printMessage "install rca0" $?

set -x
export POD_RCA0=$(kubectl get pods -n n0 -l "app=hlf-ca,release=rca0" -o jsonpath="{.items[0].metadata.name}")
kubectl wait --for=condition=Ready --timeout 180s pod/$POD_RCA0 -n n0
res=$?
set +x
printMessage "pod/$POD_RCA0" $res

sleep 30

echo "#################################"
echo "### Step 9: crypto-tlsca0"
echo "#################################"
helm install crypto-tlsca0 -n n0 -f ./releases/org0/tlsca0-cryptogen.gcp.yaml ./cryptogen
printMessage "install crypto-tlsca0" $?

set -x
kubectl wait --for=condition=complete --timeout 180s job/crypto-tlsca0-cryptogen -n n0
res=$?
set +x
printMessage "job/crypto-tlsca0-cryptogen" $res

sleep 30

echo "#################################"
echo "### Step 10: crypto-rca0"
echo "#################################"
helm install crypto-rca0 -n n0 -f ./releases/org0/rca0-cryptogen.gcp.yaml ./cryptogen
printMessage "install crypto-rca0" $?

set -x
kubectl wait --for=condition=complete --timeout 180s job/crypto-rca0-cryptogen -n n0
res=$?
set +x
printMessage "job/crypto-rca0-cryptogen" $res

sleep 5

echo "#################################"
echo "### Step 11: Create secrets"
echo "#################################"
./scripts/create-secret.rca0.sh
printMessage "create secret rca0" $?

./scripts/create-secret.rca1.sh
printMessage "create secret rca1" $?

sleep 30

echo "#################################"
echo "### Step 12: Create genesis block and channeltx"
echo "#################################"
######## post-install notes for admin0/orgadmin
######## Objective: These steps create geneis.block, and secret "genesis" and "channel.tx"
######## 1. Get the name of the pod running rca:
set -x
export POD_CLI0=$(kubectl get pods -n n0 -l "app=orgadmin,release=admin0" -o jsonpath="{.items[0].metadata.name}")
set +x
preventEmptyValue "pod unavailable" $POD_CLI0

sleep 2

######## 2. Create genesis.block / channel.tx / anchor.tx
set -x
kubectl -n n0 exec -it $POD_CLI0 -- sh -c "/var/hyperledger/bin/configtxgen -configPath /var/hyperledger/cli/configtx -profile OrgsOrdererGenesis -outputBlock /var/hyperledger/crypto-config/genesis.block -channelID ordererchannel"
res=$?
set +x
printMessage "create genesis block" $res
set -x
kubectl -n n0 exec -it $POD_CLI0 -- sh -c "/var/hyperledger/bin/configtxgen -configPath /var/hyperledger/cli/configtx -profile OrgsChannel -outputCreateChannelTx /var/hyperledger/crypto-config/channel.tx -channelID loanapp"
res=$?
set +x
printMessage "create channel.tx" $res

######## 3. Create configmap: genesis.block
kubectl -n n0 exec $POD_CLI0 -- cat /var/hyperledger/crypto-config/genesis.block > genesis.block
kubectl -n n0 create secret generic genesis --from-file=genesis=./genesis.block
printMessage "create secret genesis" $?
rm genesis.block

######## 4. Create configmap: channel.tx for org1, with namespace n1
kubectl -n n0 exec $POD_CLI0 -- cat /var/hyperledger/crypto-config/channel.tx > channel.tx
kubectl -n n1 create secret generic channeltx --from-file=channel.tx=./channel.tx
printMessage "create secret channeltx" $?
rm channel.tx

echo "#################################"
echo "### Step 13: Install orderers"
echo "#################################"

helm install o1 -f ./releases/org0/o1-hlf-ord.gcp.yaml -n n0 ./hlf-ord
sleep 3
helm install o2 -f ./releases/org0/o2-hlf-ord.gcp.yaml -n n0 ./hlf-ord
sleep 3
helm install o3 -f ./releases/org0/o3-hlf-ord.gcp.yaml -n n0 ./hlf-ord
sleep 3
helm install o4 -f ./releases/org0/o4-hlf-ord.gcp.yaml -n n0 ./hlf-ord
sleep 3
helm install o0 -f ./releases/org0/o0-hlf-ord.gcp.yaml -n n0 ./hlf-ord

set -x
export POD_O0=$(kubectl get pods -n n0 -l "app=hlf-ord,release=o0" -o jsonpath="{.items[0].metadata.name}")
kubectl wait --for=condition=Ready --timeout 180s pod/$POD_O0 -n n0
res=$?
set +x
printMessage "pod/o0-hlf-ord" $res

echo "#################################"
echo "### Step 14: Install p0o1"
echo "#################################"
helm install p0o1 -n n1 -f ./releases/org1/p0o1-hlf-peer.gcp.yaml ./hlf-peer

set -x
export POD_P0O1=$(kubectl get pods -n n1 -l "app=hlf-peer,release=p0o1" -o jsonpath="{.items[0].metadata.name}")
kubectl wait --for=condition=Ready --timeout 180s pod/$POD_P0O1 -n n1
res=$?
set +x
printMessage "pod/p0o1-hlf-peer" $res

echo "#################################"
echo "### Step 15: Install g1"
echo "#################################"
helm install g1 -n n1 -f ./releases/org1/g1-gupload.gcp.yaml ./gupload

export POD_CLI1=$(kubectl get pods --namespace n1 -l "app=orgadmin,release=admin1" -o jsonpath="{.items[0].metadata.name}")
preventEmptyValue "pod unavailable" $POD_CLI1

sleep 10s
echo "#################################"
echo "### Step 16: Bootstrap part 1"
echo "#################################"
helm install b1 -n n1 -f ./releases/org1/bootstrap-a.gcp.yaml ./hlf-operator
set -x
kubectl wait --for=condition=complete --timeout 300s job/b1-hlf-operator--bootstrap -n n1
res=$?
set +x
printMessage "job/bootstrap part1" $res

echo "#################################"
echo "### Step 17: Install chaincode"
echo "#################################"
set -x
export CCID=$(kubectl -n n1 exec $POD_CLI1 -- cat /var/hyperledger/crypto-config/channel-artifacts/packageid.txt)
res=$?
set +x
printMessage "retrieve CCID" $res

helm install eventstore -n n1 --set ccid=$CCID -f ./releases/org1/eventstore-hlf-cc.gcp.yaml ./hlf-cc
set -x
export POD_CC1=$(kubectl get pods -n n1 -l "app=hlf-cc,release=eventstore" -o jsonpath="{.items[0].metadata.name}")
kubectl wait --for=condition=Ready --timeout 180s pod/$POD_CC1 -n n1
res=$?
set +x
printMessage "pod/eventstore chaincode" $res

sleep 10s

echo "#################################"
echo "### Step 18: Bootstrap part 2"
echo "#################################"
helm install b2 -n n1 -f ./releases/org1/bootstrap-b.gcp.yaml ./hlf-operator
set -x
kubectl wait --for=condition=complete --timeout 300s job/b2-hlf-operator--bootstrap -n n1
res=$?
set +x
printMessage "job/bootstrap part2" $res

echo "#################################"
echo "### Step 19: Upload org1 - tls root certs"
echo "#################################"
export POD_CLI1=$(kubectl get pods --namespace n1 -l "app=orgadmin,release=admin1" -o jsonpath="{.items[0].metadata.name}")
preventEmptyValue "pod unavailable" $POD_CLI1

# download org1 root cert
set -x
kubectl -n n1 exec $POD_CLI1 -- cat ./Org1MSP/msp/tlscacerts/tls-ca-cert.pem > ./download/org1.net-tlscacert.pem
res=$?
set +x
printMessage "download Org1MSP/msp/tlscacerts/tls-ca-cert.pem from n1" $res

export POD_G1=$(kubectl get pods --namespace n1 -l "app=gupload,release=g1" -o jsonpath="{.items[0].metadata.name}")
preventEmptyValue "pod unavailable" $POD_G1

# send org1 root cert to 'public' folder of gupload
set -x
kubectl -n n1 cp ./download/org1.net-tlscacert.pem $POD_G1:/var/gupload/fileserver/public/ -c gupload
res=$?
set +x
printMessage "cp org1.net-tlscacert.pem to g1-gupload" $res

echo "#################################"
echo "### Step 20: Upload org0 - tls root certs"
echo "#################################"
export POD_CLI0=$(kubectl get pods --namespace n0 -l "app=orgadmin,release=admin0" -o jsonpath="{.items[0].metadata.name}")
preventEmptyValue "pod unavailable" $POD_CLI1

# download org0 root cert
set -x
kubectl -n n0 exec $POD_CLI0 -- cat ./Org0MSP/msp/tlscacerts/tls-ca-cert.pem > ./download/org0.com-tlscacert.pem
res=$?
set +x
printMessage "download Org0MSP/msp/tlscacerts/tls-ca-cert.pem from n0" $res

# send org1 root cert to 'public' folder of gupload
set -x
kubectl -n n1 cp ./download/org0.com-tlscacert.pem $POD_G1:/var/gupload/fileserver/public/ -c gupload
res=$?
set +x
printMessage "cp org0.com-tlscacert.pem to g1-gupload" $res

duration=$SECONDS
printf "${GREEN}$(($duration / 60)) minutes and $(($duration % 60)) seconds elapsed.\n\n${NC}"
